File: src-tauri/src/repo/write.rs
//! Write paths and anything gix does not yet cover cleanly: we invoke the
//! user's installed `git` binary as a subprocess. This is the single boundary
//! where commit/stash/worktree/rebase operations will live.

use anyhow::{bail, Result};
use std::process::Command;
#[cfg(windows)]
use std::os::windows::process::CommandExt;

/// Don't pop a console window for each subprocess we spawn (Windows GUI apps
/// otherwise flash a console on every `git` call).
#[cfg(windows)]
const CREATE_NO_WINDOW: u32 = 0x0800_0000;

/// A `Command` that never flashes a console window on Windows.
pub fn command(program: &str) -> Command {
    let mut c = Command::new(program);
    #[cfg(windows)]
    c.creation_flags(CREATE_NO_WINDOW);
    c
}

/// Backoff schedule for retrying a git command that lost the race for the
/// index lock (an agent committing while the user stages, or vice versa).
const LOCK_RETRY_MS: &[u64] = &[100, 300, 800, 1500];

/// Does this stderr indicate transient lock contention (safe to retry)?
/// We never delete the lock file ourselves — the other process owns it.
fn is_lock_error(stderr: &str) -> bool {
    let m = stderr.to_lowercase();
    m.contains("index.lock")
        || m.contains("another git process")
        || (m.contains("unable to create") && m.contains(".lock"))
}

fn run_git(workdir: &str, extra: &[&str], args: &[&str]) -> Result<String> {
    let out = command("git")
        .arg("-C")
        .arg(workdir)
        .args(extra)
        .args(args)
        .output()?;

    if !out.status.success() {
        bail!(
            "git {} failed: {}",
            args.join(" "),
            String::from_utf8_lossy(&out.stderr).trim()
        );
    }

    Ok(String::from_utf8_lossy(&out.stdout).into_owned())
}

fn git_with_retry(workdir: &str, extra: &[&str], args: &[&str]) -> Result<String> {
    let mut attempt = 0;
    loop {
        match run_git(workdir, extra, args) {
            Ok(out) => return Ok(out),
            Err(e) if attempt < LOCK_RETRY_MS.len() && is_lock_error(&e.to_string()) => {
                std::thread::sleep(std::time::Duration::from_millis(LOCK_RETRY_MS[attempt]));
                attempt += 1;
            }
            Err(e) => return Err(e),
        }
    }
}

/// Run `git -C <workdir> <args...>` and return stdout on success.
/// Retries briefly on index-lock contention.
pub fn git(workdir: &str, args: &[&str]) -> Result<String> {
    git_with_retry(workdir, &[], args)
}

/// Read-path variant: `--no-optional-locks` stops git from taking its
/// opportunistic locks (e.g. the status untracked-cache refresh), so reads
/// can never collide with an agent mid-commit. Designed exactly for tools
/// like Grove that run status in the background.
pub fn git_read(workdir: &str, args: &[&str]) -> Result<String> {
    git_with_retry(workdir, &["--no-optional-locks"], args)
}

/// Stage one file (`git add`).
pub fn stage(path: &str, file: &str) -> Result<()> {
    let dir = super::read::workdir_of(path)?;
    git(&dir, &["add", "--", file])?;
    Ok(())
}

/// Unstage one file (`git restore --staged`).
pub fn unstage(path: &str, file: &str) -> Result<()> {
    let dir = super::read::workdir_of(path)?;
    git(&dir, &["restore", "--staged", "--", file])?;
    Ok(())
}

/// Stage everything (`git add -A`).
pub fn stage_all(path: &str) -> Result<()> {
    let dir = super::read::workdir_of(path)?;
    git(&dir, &["add", "-A"])?;
    Ok(())
}

/// Unstage everything (`git reset`).
pub fn unstage_all(path: &str) -> Result<()> {
    let dir = super::read::workdir_of(path)?;
    git(&dir, &["reset"])?;
    Ok(())
}

/// Commit the staged changes with `message`.
pub fn commit(path: &str, message: &str) -> Result<String> {
    let dir = super::read::workdir_of(path)?;
    git(&dir, &["commit", "-m", message])
}

#[cfg(test)]
mod tests {
    use super::is_lock_error;

    #[test]
    fn lock_errors_are_recognised() {
        assert!(is_lock_error(
            "fatal: Unable to create 'C:/r/.git/index.lock': File exists.\n\nAnother git process seems to be running"
        ));
        assert!(is_lock_error("error: could not lock config file .git/config: index.lock held"));
        assert!(!is_lock_error("fatal: not a git repository"));
        assert!(!is_lock_error("error: pathspec 'foo' did not match any file(s)"));
    }
}

/// Clone `url` into `dest` (a directory that must not already exist).
pub fn clone(url: &str, dest: &str) -> Result<()> {
    let out = command("git").args(["clone", url, dest]).output()?;
    if !out.status.success() {
        bail!(
            "git clone failed: {}",
            String::from_utf8_lossy(&out.stderr).trim()
        );
    }
    Ok(())
}

File: engine/src/write.ts
/**
 * Write paths: the mutations Grove performs on a repository.
 * Every one of these goes through `git()`, so every one inherits the
 * index-lock retry. Ported from `src-tauri/src/repo/write.rs`.
 */

import { mkdirSync, existsSync } from 'node:fs';
import { join } from 'node:path';

import { workdirOf } from './discover.ts';
import { git, gitBare } from './git.ts';

/** Stage one file (`git add`). */
export async function stage(path: string, file: string): Promise<void> {
  await git(workdirOf(path), ['add', '--', file]);
}

/** Unstage one file (`git restore --staged`). */
export async function unstage(path: string, file: string): Promise<void> {
  await git(workdirOf(path), ['restore', '--staged', '--', file]);
}

/** Stage everything (`git add -A`). */
export async function stageAll(path: string): Promise<void> {
  await git(workdirOf(path), ['add', '-A']);
}

/** Unstage everything (`git reset`). */
export async function unstageAll(path: string): Promise<void> {
  await git(workdirOf(path), ['reset']);
}

/** Commit the staged changes with `message`. */
export function commit(path: string, message: string): Promise<string> {
  return git(workdirOf(path), ['commit', '-m', message]);
}

/** Clone `url` into `dest` (a directory that must not already exist). */
export async function clone(url: string, dest: string): Promise<void> {
  await gitBare(['clone', url, dest]);
}

function homeDir(): string {
  return process.env['USERPROFILE'] || process.env['HOME'] || '.';
}

/**
 * Clone `url` into `~/GroveRepos/<name>` and return the local path.
 * Ported from the `clone_repo` command in `lib.rs`.
 */
export async function cloneToGroveRepos(url: string): Promise<string> {
  const name = url.replace(/\/+$/, '').split('/').pop()?.replace(/\.git$/, '') ?? '';
  if (name === '') throw new Error('could not derive a repo name from the URL');

  const parent = join(homeDir(), 'GroveRepos');
  mkdirSync(parent, { recursive: true });

  const dest = join(parent, name);
  if (existsSync(dest)) throw new Error(`${dest} already exists`);

  await clone(url, dest);
  return dest;
}

/**
 * Best-effort `git commit-graph write --reachable`, once per repo per session,
 * in the background. The commit-graph file gives the git CLI generation-number
 * traversal, which speeds up every log/walk on large histories.
 */
const commitGraphDone = new Set<string>();

export function maybeWriteCommitGraph(path: string): void {
  if (commitGraphDone.has(path)) return;
  commitGraphDone.add(path);
  // Deliberately not awaited: this is a warm-up, not a dependency.
  void git(path, ['commit-graph', 'write', '--reachable']).catch(() => {});
}
